File: src/tasks.js
"use strict";

/**
 * @param {unknown} raw
 * @returns {{ project: object | null, columns: object[], allTasks: object[], archived: object[], planned: object[] }}
 */
function parseBoardResponse(raw) {
  const data = raw?.data ?? raw;
  if (!data) {
    return { project: null, columns: [], allTasks: [], archived: [], planned: [] };
  }

  const columns = Array.isArray(data.columns) ? data.columns : [];
  const allTasks = [];
  for (const col of columns) {
    for (const task of col.tasks || []) {
      allTasks.push({
        ...task,
        status: task.status || col.id || col.slug,
        statusName: col.name || task.status,
      });
    }
  }

  const archived = (data.archivedTasks || []).map((t) => ({ ...t, status: t.status || "archived" }));
  const planned = (data.plannedTasks || []).map((t) => ({ ...t, status: t.status || "planned" }));

  return {
    project: {
      id: data.id,
      name: data.name,
      slug: data.slug,
      workspaceId: data.workspaceId,
    },
    columns,
    allTasks: [...allTasks, ...archived, ...planned],
    archived,
    planned,
  };
}

/**
 * @param {object} task
 * @param {string} [projectSlug]
 */
function taskKey(task, projectSlug) {
  const slug = String(projectSlug || "TASK").toUpperCase();
  const num = task.number != null ? task.number : "?";
  return `${slug}-${num}`;
}

const PRIORITY_LABELS = {
  urgent: "Urgent",
  high: "High",
  medium: "Medium",
  low: "Low",
  "no-priority": "None",
};

const STATUS_COLORS = {
  "to-do": "#6b7280",
  "in-progress": "#2563eb",
  "in-review": "#7c3aed",
  done: "#16a34a",
  archived: "#9ca3af",
  planned: "#d97706",
};

module.exports = {
  parseBoardResponse,
  taskKey,
  PRIORITY_LABELS,
  STATUS_COLORS,
};

File: neovim/lua/kaneo/tasks.lua
local M = {}

function M.task_key(task, project_slug)
  local slug = string.upper(project_slug or "TASK")
  local num = task.number ~= nil and task.number or "?"
  return slug .. "-" .. tostring(num)
end

function M.parse_board_response(raw)
  local data = raw
  if type(raw) == "table" and raw.data then
    data = raw.data
  end
  if not data then
    return { project = nil, columns = {}, all_tasks = {} }
  end

  local columns = data.columns or {}
  local all_tasks = {}
  for _, col in ipairs(columns) do
    for _, task in ipairs(col.tasks or {}) do
      table.insert(all_tasks, vim.tbl_extend("force", task, {
        status = task.status or col.id or col.slug,
        statusName = col.name or task.status,
      }))
    end
  end

  for _, t in ipairs(data.archivedTasks or {}) do
    table.insert(all_tasks, vim.tbl_extend("force", t, { status = t.status or "archived" }))
  end
  for _, t in ipairs(data.plannedTasks or {}) do
    table.insert(all_tasks, vim.tbl_extend("force", t, { status = t.status or "planned" }))
  end

  return {
    project = {
      id = data.id,
      name = data.name,
      slug = data.slug,
      workspaceId = data.workspaceId,
    },
    columns = columns,
    all_tasks = all_tasks,
  }
end

M.PRIORITY_LABELS = {
  urgent = "Urgent",
  high = "High",
  medium = "Medium",
  low = "Low",
  ["no-priority"] = "None",
}

function M.format_date_short(iso)
  if not iso or iso == "" then
    return ""
  end
  return tostring(iso):sub(1, 10)
end

return M
